### PR TITLE
[Reg 2.074.1] Fix Issue  19569 Fixed overload resolution giving priority to covariants when should be ambiguous.

### PR DESCRIPTION
**Pull request by ghost**
Culprit https://github.com/dlang/dmd/commit/832031adddadc3b9b8a3bca94bd30be1393b570a#diff-4a5b098dcf557585b90b74a8c638b0c9R2566

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]
Looks fine to me, @RazvanN7 would know more about the intricacies.

File: src/dmd/dtemplate.d [hunk 1]
```diff
@@ -2505,8 +2505,16 @@ void functionResolve(Match* m, Dsymbol dstart, Loc loc, Scope* sc, Objects* tiar
             if (!m.lastf.type.equals(fd.type))
             {
                 //printf("cov: %d %d\n", m.lastf.type.covariant(fd.type), fd.type.covariant(m.lastf.type));
-                if (m.lastf.type.covariant(fd.type) == 1) goto LlastIsBetter;
-                if (fd.type.covariant(m.lastf.type) == 1) goto LfIsBetter;
+                int lastCovariant = m.lastf.type.covariant(fd.type);
+                int firstCovariant = fd.type.covariant(m.lastf.type);
```
Comment (jacob-carlborg) [inline src/dmd/dtemplate.d]: `const` for both variables.

File: src/dmd/dtemplate.d [hunk 2]
```diff
@@ -2505,8 +2505,16 @@ void functionResolve(Match* m, Dsymbol dstart, Loc loc, Scope* sc, Objects* tiar
...
+
+                if (lastCovariant == 1 || lastCovariant == 2) {
```
Comment (jacob-carlborg) [inline src/dmd/dtemplate.d]: Curly braces on their own lines.

**Comment by RazvanN7** [comment]
I'm not sure this is the right solution. I haven't found anything about overloading on function attributes in the spec, but currently it seems that the compiler prefers the "better" function which is the one annotated with an attribute.

**Comment by look-at-me** [comment]
How does it determine it is better? Right now it just picks the one with an attribute. If there are two with attributes only then does the call become ambiguous. It never picked the one that was better suited to the current scope. It doesn't use the current scope for selection, only after it is selected does it complain.

```D
import core.stdc.stdio;

void foo() @nogc  { printf("foo @nogc"); }
void foo() nothrow { printf("foo nothrow"); }

void main() @nogc {
    foo(); // error ambiguous
}
```
https://run.dlang.io/is/fpwiwG

```D
import core.stdc.stdio;

void foo() nothrow { printf("foo nothrow"); }
void foo() { printf("foo"); }

void main() @nogc {
   	(cast(void function() @nogc)&foo)(); // prints `foo nothrow`
}
```
https://run.dlang.io/is/bbJpNp

**Comment by RazvanN7** [comment]
I see. It picks the first one that is defined. My bad; I thought it picked the qualified one because it offers stronger guarantees.

**Review by RazvanN7 — APPROVED** [review]

**Comment by Basile-z** [comment]
This fix could go in stable, as it's a regression. @look-at-me, are you available to rebase your work quick ? IIRC I've read somewhere that the point release should come fast.

**Comment by RazvanN7** [comment]
I think it would be wiser if we would just merge it to master and see if there are any side effects that we didn't think of.

**Pull request merged**